### PR TITLE
[processing] Handle case of folder of shapefiles in build vrt alg

### PR DESCRIPTION
**Pull request by volaya**
#fixes 21519

## Description

If a shapefile (i.e /my/data/points.shp) was loaded by loading a directory of shapefiles, its source wont be the filepath, but instead a uri-like thing such as /my/data|layername=points.

Although that file can be read by the external tool that creates the virtual vector, that source string won't work. I added a patch to try to resolve the uri-like source to a regular filepath.

It's just a patch for this tool, but this might affect other tools as well (basically, all the ones with external apps, like SAGA, etc), which wont work with shpafiles loaded as part of a folder. 

Ideally, there should be a method to resolve the source to a valid filepath if possible, but I couldnt find it. I am making the PR in any case, to get some discussion about this.

I will be happy to improve this if someone points me in the direction of such method of workflow, in case it exists.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: python/plugins/processing/algs/qgis/Datasources2Vrt.py [hunk 1]
```diff
@@ -125,11 +127,17 @@ def mergeDataSources2Vrt(self, dataSources, outFile, union=False, relative=False
...
-            inFile = layer.source()
+            inFile = layer.source()            
             srcDS = ogr.Open(inFile, 0)
             if srcDS is None:
-                raise QgsProcessingException(
-                    self.tr('Invalid datasource: {}'.format(inFile)))
+                #might be a shapefile loaded as a directory of shapefiles. We try to resolve the uri to a filepath
+                sourceParts = QgsProviderRegistry.instance().decodeUri('ogr', inFile)
+                if sourceParts.get("layerName"):
+                    inFile = os.path.join(sourceParts.get("path"), sourceParts.get("layerName") + ".shp")
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/Datasources2Vrt.py]: Won't this break if a geopackage layer (or similar) was loaded? In that case the layerName parameter will be present, but the file name generated here will be broken.

This is really quite tricky to handle -- I think we'd need a check in here to first see if sourceParts['path'] is actually a directory (and if not, there's nothing required). Then, if it IS a directory, we'd need to check inside that path for files which start with the layername and have known vector data format extensions (because the same situation may arise for say KML files), and use that to determine the actual name of the file. Or alternatively, maybe we could use OGR api to try to load the file using just the path and layername, and let it run whatever internal logic it has to resolve the layername to a file, and then retrieve this actual file name from the OGR layer...

**Comment by nyalldawson** [comment]
> Ideally, there should be a method to resolve the source to a valid filepath if possible, but I couldnt find it. I am making the PR in any case, to get some discussion about this.

Agreed - we need a central method for handling this. Maybe in QgsOgrUtils.cpp?

File: python/plugins/processing/algs/qgis/Datasources2Vrt.py (hunk 1 above)
Comment (volaya) [reply python/plugins/processing/algs/qgis/Datasources2Vrt.py]: It wont actually break anything, since gpkgs dont work now. Any layer with a source that is not an actual filepath cannot be used with that algorithm at the moment. However, I agree that the trick i added is not very general and only fixes the case of shapefiles.

I improved it a bit and put it in a separate function in the processing.tools.vector module. From there, it can be moved to C++ if someone takes care of that.

Still, some cases are not handled. It works now for gpkgs, but as long as the layer is the only one in the gpkg file. 

In short, it's an improvement over the previous status of the algorithm, but it wont work with all input layers, like other Processing algorithms do. Other algorithms might be affected of this issue, so it's worth checking and maybe adding a similar mechanism.

**Comment by nyalldawson** [comment]
@volaya I've asked here https://lists.osgeo.org/pipermail/gdal-dev/2019-June/050380.html to see if there's a cleaner way we can approach this

**Comment by volaya** [comment]
cool. Good idea. Let's wait to see what they say

**Comment by m-kuhn** [comment]
Feedback @rouault 

> > My question is -- GetName() here returns just the file base name (i.e.
> > "lines" for "/path/lines.shp"). Is there anyway to retrieve the actual
> > layer file path for these?
> 
> No, for most formats, there is no direct mapping from a layer to file(s) 
> containing layer data, so there is nothing in the OGR API for that. So you 
> have to use the knowledge that this is opened by the shapefile or mapinfo + 
> that the datasource name is a directory, and then you can form a filename from 
> that and the layer name.

File: python/plugins/processing/tools/vector.py [hunk 2]
```diff
@@ -111,3 +114,20 @@ def checkMinDistance(point, index, distance, points):
...
+def resolveSourceToFilepath(source):
+    #Tries to resolve a uri-like source (i.e /path/|layername=points) to a valid filepath (/path/points.shp)
+    sourceParts = QgsProviderRegistry.instance().decodeUri('ogr', source)
+    folder = sourceParts.get("path")
+    layerName = sourceParts.get("layerName")
+    if layerName:
+        if os.path.exists(folder):
+            if os.path.isdir(folder):
+                for ext in QgsVectorFileWriter.supportedFormatExtensions():
+                    f = os.path.join(folder, "%s.%s" % (layerName, ext))
```
Comment (m-kuhn) [inline python/plugins/processing/tools/vector.py]:
```suggestion
                    f = os.path.join(folder, "{basename}.{extension}".format(basename=layerName, extension=ext))
```

Just because ;)
